wave_lang/kernel/wave/scheduling/multi_buffering.py

```diff
@@ -41,104 +88,33 @@ def multi_buffer(trace: CapturedTrace):
     reduction_axis = get_custom(reductions[0]).axis
 
     # Find reads and writes operating on shared memory
-    reads = []
-    writes = []
+    reads_and_writes = list()
     for node in trace.get_subgraph(get_custom(reductions[0]).subgraph_name).nodes:
         custom = get_custom(node)
         if (
             isinstance(custom, Read | Write)
             and custom.memory_type.address_space == SHARED_ADDRESS_SPACE
         ):
-            if isinstance(custom, Read):
-                reads.append(custom)
-            elif isinstance(custom, Write):
-                writes.append(custom)
+            reads_and_writes.append(custom)
 
     # Partition reads and writes by memory location
-    memory_to_reads = _partition_by_memory(reads)
-    memory_to_writes = _partition_by_memory(writes)
-
+    memory_to_shared_memory_op = _partition_by_memory(reads_and_writes)
     # Perform multi buffering for all collected memory locations
-    for memory_location in set(memory_to_reads.keys()) | set(memory_to_writes.keys()):
-        read_nodes = memory_to_reads.get(memory_location, [])
-        write_nodes = memory_to_writes.get(memory_location, [])
-
-        _multi_buffer_memory_location(
-            trace, memory_location, read_nodes, write_nodes, reduction_axis, 2
+    for memory_location in set(memory_to_shared_memory_op):
+        _increase_dimensions_for_multibuffering(
+            original_buffer=memory_location,
+            reduction_axis=reduction_axis,
+            buffer_count=multi_buffer_count,
         )
 
 
-def _multi_buffer_memory_location(
-    trace: CapturedTrace,
-    original_buffer: CustomOp,
-    read_nodes: list[Read],
-    write_nodes: list[Write],
-    reduction_axis: IndexSymbol,
-    buffer_count: int,
+def _increase_dimensions_for_multibuffering(
```

Comment:
`_increase_dimensions_for_multibuffering` currently multiplies each non-reduction dim by `buffer_count` which will cause issues if we have more than 1, I think we should multiply only one dim.